app/validators/questionnaire_schema.py

```diff
@@ -384,3 +339,76 @@ def get_first_answer_in_block(self, block_id):
     def _get_path_id(self, path):
         return jp.match1(path + ".id", self.schema)
+
+    def get_numeric_range_values(self, answer, answer_ranges):
+        min_value = answer.get("minimum", {}).get("value", {})
+        max_value = answer.get("maximum", {}).get("value", {})
+        min_referred = (
+            min_value.get("identifier") if isinstance(min_value, dict) else None
+        )
+        max_referred = (
+            max_value.get("identifier") if isinstance(max_value, dict) else None
+        )
+
+        exclusive = answer.get("exclusive", False)
+        decimal_places = answer.get("decimal_places", 0)
+
+        return {
+            "min": self.get_answer_minimum(
+                min_value, decimal_places, exclusive, answer_ranges
+            ),
+            "max": self.get_answer_maximum(
+                max_value, decimal_places, exclusive, answer_ranges
+            ),
+            "decimal_places": decimal_places,
+            "min_referred": min_referred,
+            "max_referred": max_referred,
+            "default": answer.get("default"),
+        }
+
+    def get_answer_minimum(
+        self, defined_minimum, decimal_places, exclusive, answer_ranges
+    ):
+        minimum_value = self.get_numeric_value(defined_minimum, 0, answer_ranges)
+        if exclusive:
+            return minimum_value + (1 / 10 ** decimal_places)
+        return minimum_value
+
+    def get_answer_maximum(
+        self, defined_maximum, decimal_places, exclusive, answer_ranges
+    ):
+        maximum_value = self.get_numeric_value(
+            defined_maximum, MAX_NUMBER, answer_ranges
+        )
+        if exclusive:
+            return maximum_value - (1 / 10 ** decimal_places)
+        return maximum_value
+
+    def get_numeric_value(self, defined_value, system_default, answer_ranges):
+        if not isinstance(defined_value, dict):
+            return defined_value
+        if "source" in defined_value:
+            referred_answer = None
+            if defined_value["source"] == "answers":
+                referred_answer = answer_ranges.get(defined_value["identifier"])
+
+            elif defined_value["source"] == "calculated_summary":
+                min_calculated_summary_block = self.get_block(
+                    defined_value["identifier"]
+                )
+                answers_to_calculate = min_calculated_summary_block["calculation"][
+                    "answers_to_calculate"
+                ]
+
+                for answer_id in answers_to_calculate:
+                    referred_answer = answer_ranges.get(answer_id)
+                    if referred_answer is None:
+                        return None
+
+            if referred_answer is None:
+                # Referred answer is not valid (picked up by _validate_referred_numeric_answer)
+                return None
+            if referred_answer.get("default") is not None:
+                return system_default
+
+        return system_default
```

Comment:
I think we need to reduce the cognitive complexity. It could be made into 2 methods like so:

```suggestion
    @staticmethod
    def _get_numeric_value_for_value_source(
        *, value_source, defined_value, answer_ranges
    ):
        referred_answer = None
        if value_source == "answers":
            referred_answer = answer_ranges.get(defined_value["identifier"])

        elif value_source == "calculated_summary":
            answers_to_calculate = []

            for answer_id in answers_to_calculate:
                referred_answer = answer_ranges.get(answer_id)
                if referred_answer is None:
                    return None

        return referred_answer

    def get_numeric_value(self, defined_value, system_default, answer_ranges):
        if not isinstance(defined_value, dict):
            return defined_value

        if source := defined_value.get("source"):
            referred_answer = self._get_numeric_value_for_value_source(
                value_source=source,
                defined_value=defined_value,
                answer_ranges=answer_ranges,
            )

            # Referred answer is not valid (picked up by _validate_referred_numeric_answer)
            if not referred_answer:
                return None

        return system_default
```

Reply:
Have updated this now 👍